packages/edit-navigation/src/components/header/use-menu-locations.js

```diff
@@ -0,0 +1,68 @@
+/**
+ * WordPress dependencies
+ */
+import { useState, useEffect } from '@wordpress/element';
+import apiFetch from '@wordpress/api-fetch';
+import { useDispatch } from '@wordpress/data';
+
+export default function useMenuLocations() {
+	const [ menuLocationsByName, setMenuLocationsByName ] = useState( null );
+
+	useEffect( () => {
+		const fetchMenuLocationsByName = async () => {
+			const newMenuLocationsByName = await apiFetch( {
+				method: 'GET',
+				path: '/__experimental/menu-locations',
+			} );
+			setMenuLocationsByName( newMenuLocationsByName );
+		};
+		fetchMenuLocationsByName();
+	}, [] );
+
+	const { saveMenu } = useDispatch( 'core' );
+
+	const assignMenuToLocation = async ( locationName, newMenuId ) => {
```

Comment:
Maybe we'll want to memoize this function, so that the reference to the setter is always identical.

Reply:
My understanding is that that's only necessary if we're passing `assignMenuToLocation` as a prop. Is that not right?

Reply:
No, it doesn't :). Just that if we ever want to use it in other components, the developer is then expected to know that `assignMenuToLocation` is not gonna be identical between re-renders. If the parent component uses it in `useEffect` or other hooks, then they are also expected to know that it will cause effect the run between every renders.
To me, it's just an unnecessary mental overhead I have to care about as a developer. Furthermore, this hook returns a `useState`-like API, which could be naturally considered to match the behavior of it as well, which is that the setter function shouldn't change between re-renders.

Reply:
Fair enough! The point about callers expecting this hook to act like `useState` is a good one. Done in a82a97c3370321bebdb52bf919e10df0dbbf05f2.

I see `useMemo()` and `useCallback()` in our codebase a lot but rarely do I see `memo()` or `PureComponent` which makes me wonder if we're actually hurting performance from unnecessary memoization 😅